lib/Targets/AIETargetCDO.cpp

```diff
@@ -0,0 +1,1084 @@
+//===- AIETargetCDO.cpp -----------------------------------------*- C++ -*-===//
+//
+// Copyright (C) 2023, Advanced Micro Devices, Inc. All rights reserved.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#include "AIETargetShared.h"
+#include "AIETargets.h"
+
+#include "aie/Dialect/AIE/IR/AIEDialect.h"
+#include "aie/Dialect/AIEX/IR/AIEXDialect.h"
+
+#include "mlir/IR/Attributes.h"
+#include "mlir/IR/IRMapping.h"
+#include "mlir/Pass/Pass.h"
+#include "mlir/Tools/mlir-translate/MlirTranslateMain.h"
+
+#include "llvm/ADT/StringExtras.h"
+#include "llvm/IR/Module.h"
+
+using namespace mlir;
+using namespace xilinx;
+using namespace xilinx::AIE;
+using namespace xilinx::AIEX;
+
+namespace xilinx {
+namespace AIE {
+
+int maxCol{0}, minCol{0};
+static std::string numColsStr() { return std::to_string(1 + maxCol - minCol); }
+
+static std::string cdoGenFileHeader() {
+  std::string str;
+  llvm::raw_string_ostream rss(str);
+  rss << "/********************************************* Disclaimer "
+         "*********************************************/\n"
+      << "/* This file is generated by aie-translate. */\n"
+      << "/* Changes to this file may cause incorrect behavior. */\n"
+      << "\n"
+      << "/************************** Constants/Macros "
+         "*****************************/\n"
+      << "#define HW_GEN                   XAIE_DEV_GEN_AIEML\n"
+      << "#define XAIE_NUM_ROWS            6\n"
+      << "#define XAIE_NUM_COLS            5\n"
+      << "#define XAIE_BASE_ADDR           0x40000000\n"
+      << "#define XAIE_COL_SHIFT           25\n"
+      << "#define XAIE_ROW_SHIFT           20\n"
+      << "#define XAIE_SHIM_ROW            0\n"
+      << "#define XAIE_MEM_TILE_ROW_START  1\n"
+      << "#define XAIE_MEM_TILE_NUM_ROWS   1\n"
+      << "#define XAIE_AIE_TILE_ROW_START  2\n"
+      << "#define XAIE_AIE_TILE_NUM_ROWS   4\n"
+      << "#define FOR_WRITE                0\n"
+      << "#define FOR_READ                 1\n"
+      << "#define XAIE_PARTITION_BASE_ADDR 0x0\n"
+      << "\n"
+      << "/***************************** Includes "
+         "*********************************/\n"
+      << "//#include <fstream>\n"
+      << "extern \"C\"\n"
+      << "{\n"
+      << "  #include <xaiengine.h>\n"
+      << "}\n"
+      << "//#include \"adf/adf_api/AIEControlConfig.h\"\n\n"
+      << "#define __mlir_aie_try(x) x\n"
+      << "static XAie_DmaDimDesc *__mlir_aie_alloc_dim_desc(size_t ndims) {\n"
+      << "  XAie_DmaDimDesc *ret = NULL;\n"
+      << "  ret = (XAie_DmaDimDesc *)calloc(sizeof(XAie_DmaDimDesc), ndims);\n"
+      << "  if(NULL == ret) {\n"
+      << "    fprintf(stderr, \"Allocating DmaDimDesc failed.\\n\");\n"
+      << "  }\n"
+      << "  return ret;\n"
+      << "}\n";
```

Comment:
Why not using raw string literals?
https://en.cppreference.com/w/cpp/language/string_literal

Reply:
for these long unparameterized ones I agree. For the parameterized ones, I would consider taking on https://github.com/fmtlib/fmt which implements both [C++20 std::format](https://en.cppreference.com/w/cpp/utility/format) and [C++23 std::print](https://en.cppreference.com/w/cpp/io/print) (so eventually we'll be able to just remove the dep easily).